### DISCUSSION TITLE
How to properly include native librairies?

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#12962] [Q&A]
Hi, I'm currently trying to add audio support for all platforms in my game, I have implemented the necessary elements and managed to find a way to make it work in most situations on desktops platforms.
This is the relevant content of my core csproj file :

```csproj
<ItemGroup Condition="$(IncludeWindowsNativeLibs) == 'true'">
    <_NativeFile Include="$(MSBuildThisFileDirectory)native/win-x86/openal32.dll" Condition="'$(RuntimeIdentifier)' != 'win-x64'">
        <Dir Condition="$(IDELaunch) == 'true'">runtimes/win-x86/native/</Dir>
        <Dir Condition="$(IDELaunch) != 'true'"></Dir>
    </_NativeFile>
    <_NativeFile Include="$(MSBuildThisFileDirectory)native/win-x64/openal32.dll" Condition="'$(RuntimeIdentifier)' != 'win-x86'">
        <Dir Condition="$(IDELaunch) == 'true'">runtimes/win-x64/native/</Dir>
        <Dir Condition="$(IDELaunch) != 'true'"></Dir>
    </_NativeFile>

    <None Include="@(_NativeFile)">
        <Link>%(Dir)%(Filename)%(Extension)</Link>
        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
        <Pack>true</Pack>
    </None>
</ItemGroup>
```

However, i'm unsure that this is the proper way to place the native librairies properly inside the `/runtimes/{Runtime Identifier}/native` folder in the case of an IDE launch or `{Runtime Identifier}/` in the case of a publishing my game.

Not only that, android doesn't seem to have the necessary library by default therefore i need to figure out a way to include another custom built shared library for this platform too somehow.
However, the way native librairies are included in the android app seem to be different.

Would it be possible to provide tips on how to properly include native librairies with my cross-platform Avalonia app?

Update : Managed to find the proper way to include native libs for the android app, turns out you can use the same mean as in xamarin to include native librairies there

```csproj
<ItemGroup Condition="$(IncludeAndroidNativeLibs) == 'true'">
    <AndroidNativeLibrary Include="../ColorTiles/native/android/libopenal.so" Abi="armeabi-v7a" />
    <AndroidNativeLibrary Include="../ColorTiles/native/android/libopenal.so" Abi="arm64-v8a" />
    <AndroidNativeLibrary Include="../ColorTiles/native/android/libopenal.so" Abi="x86" />
    <AndroidNativeLibrary Include="../ColorTiles/native/android/libopenal.so" Abi="x86_64" />
</ItemGroup>
```

**Comment by thevortexcloud** [comment]
Generally speaking, they don't actually need to be in that folder. Instead they just need to be on the search path with a matching name. Which folders are searched depends on which OS it runs on. But on all desktop OSes it will try to load them from the same directory your main application binary is located in. On for Linux for example, it will also check all the system library folders for them. EG `/lib/`. Although the exact name of the file may vary by OS if you are relying on the system provided version. Also on Linux you should be using the system library whenever possible as it's likely the version you provide wont work anyway if it has been built against a newer version of LibC.

For android specifically however I can't say how it works.

**Reply by Mrcubix** [reply]
1. Placing the dll next to the exe app does not work here, it won't be found by .NET's `NativeLibrary.Load()`
2. There are no issues on linux since it's present on most distros by default
3. I'm using OpenAL-soft's implementation, which is known to be cross-platform and works on 2 differently based distros i've tried so far.

I'm mainly looking for a way that work similarly to how how SkiaSharp does it with libSkiaSharp.dll

**Reply by thevortexcloud** [reply]
I have never used `NativeLibrary` directly, instead I normally just use `DllImport`/`LibraryImport` but in any case it looks like it can do exactly what I described with an option to enable it.

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary.load?view=net-7.0#system-runtime-interopservices-nativelibrary-load(system-string-system-reflection-assembly-system-nullable((system-runtime-interopservices-dllimportsearchpath)))

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportsearchpath?view=net-7.0

That said if you want to do the exact same way as Skiasharp then I am not 100% sure how it's done. But I _think_ it has something to do with a targets file. EG

https://github.com/mono/SkiaSharp/blob/11ad3645be6a1d20576ed1a47f00a9a5667a4006/binding/IncludeNativeAssets.HarfBuzzSharp.targets

In this case it looks like you need `AndroidNativeLibrary` somewhere.

**Reply by Mrcubix** [reply]
DllImport is just using NativeLibrary in the background.
I cannot set OpenAL's search path, if you read the doc more you'll find out you're not supposed to load a lib or set a resolver for another assembly.
I'm already using what they provide in their target file by def but not getting the same result.

I already mentionned that i found a way for android in the update.
